console/network/src/lib.rs

```
@@ -251,7 +253,11 @@ pub trait Network:
     #[cfg(any(test, feature = "test", feature = "test_consensus_heights"))]
     fn CONSENSUS_VERSION_HEIGHTS() -> &'static [(ConsensusVersion, u32); NUM_CONSENSUS_VERSIONS] {
         // NOTE: this function may panic, as it is only called during startup.
-        CONSENSUS_VERSION_HEIGHTS.get_or_init(load_test_consensus_heights)
+        #[cfg(not(feature = "wasm"))]
+        let consensus_version_heights = CONSENSUS_VERSION_HEIGHTS.get_or_init(load_test_consensus_heights);
+        #[cfg(feature = "wasm")]
+        let consensus_version_heights = CONSENSUS_VERSION_HEIGHTS.get_or_init(|| load_test_consensus_heights(None));
```

To confirm my understanding. In a WASM context, if this is called before `set_consensus_version_test_heights`, `CONSENSUS_VERSION_HEIGHTS` will be set and `set_consensus_version_test_heights` will not work right?

---

An alternate design could be to expect that this is always set in a WASM context.

---

@d0cd when the SDK starts up, it initializes a single `wasm` instance that any SnarkVM code compiled to `wasm` lives within. Thus anything static gated behind a `OnceLock` or any other object that enables one-time interior immutability creates a static memory location there.

Thus whoever is the first caller of  `CONSENSUS_VERSION_HEIGHTS.get_or_init` sets it forevermore. 

> To confirm my understanding. In a WASM context, if this is called before set_consensus_version_test_heights, CONSENSUS_VERSION_HEIGHTS will be set and set_consensus_version_test_heights will not work right?

Correct. Which is by design. The design in the SDK is to mimic the behavior with the `envar` by setting test heights if they're set by the environment variable. However this environment variable NEEDS to be set in JS and then passed in externally to initialize the static variable.

[This file specifically](https://github.com/ProvableHQ/sdk/pull/1093/files#diff-b2c9e12afcd77c0d153af2fc77a00c93fef1390a5efe9ee2450bc7a700208422R2-R13) in the SDK is how it's proposed to be done. Basically, if you're in a node context, it reads the envar, and automatically sets what's in your envar into `wasm`.  Otherwise it will initialize these heights with the default heights in the `mainnet`/`testnet`/`canary` traits when executions are run.

> An alternate design could be to expect that this is always set in a WASM context.

This puts too much burden on users and outside SDKs to manage this and is extremely error prone for people who don't know how to manage this correctly.

---

Another suggestion: `set_consensus_version_test_heights` should return an error if `CONSENSUS_VERSION_HEIGHTS` was set already. That means `set_consensus_version_test_heights` should not use `get_or_init` but https://doc.rust-lang.org/beta/std/cell/struct.OnceCell.html#method.set

---

We should probably rename this function then as I'd prefer it return the Consensus Heights so the JS users are able to see what they are!

[Example](https://github.com/ProvableHQ/sdk/pull/1093/files#diff-e0680abe1fbb50813cd1649d154a4edc9e6c907d7e91a8e3e765593aa2fe636bR39-R45)